core/services/gateway/handlers/common/callback.go

```diff
@@ -0,0 +1,43 @@
+package common
+
+import (
+	"context"
+	"errors"
+	"sync/atomic"
+
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/handlers"
+)
+
+type Callback struct {
+	ch chan handlers.UserCallbackPayload
+
+	sent       atomic.Bool
+	waitCalled atomic.Bool
+}
+
+func (c *Callback) SendResponse(payload handlers.UserCallbackPayload) error {
+	if !c.sent.CompareAndSwap(false, true) {
+		return errors.New("response already sent: each callback can only be used once")
+	}
+	// The channel is initialized with a buffer size of 1,
+	// so this send will not block.
+	c.ch <- payload
+	return nil
+}
+
+func (c *Callback) Wait(ctx context.Context) (handlers.UserCallbackPayload, error) {
+	if !c.waitCalled.CompareAndSwap(false, true) {
+		return handlers.UserCallbackPayload{}, errors.New("Wait can only be called once per Callback instance")
```

Comment:
Is there any risk that in the first call the context will be cancelled so we don't read from channel and then subsequent calls fail here? It's probably OK if there's always a single caller to Wait().